File: js/packages/browser/src/page.js
import { smartTruncate } from "./util.js";

/** Typed page markdown payload for agent handoffs. */
export class PageMarkdown {
  constructor({
    url = "",
    title = "",
    markdown = "",
    excerpt = "",
    text = "",
    links = [],
    fetchedAt = "",
    format = "markdown",
    blocked = false,
    error = "",
    markdownChars = 0,
    success = true,
  } = {}) {
    this.url = url;
    this.title = title;
    this.markdown = markdown;
    this.excerpt = excerpt || makeExcerpt(markdown || text);
    this.text = text;
    this.links = Array.isArray(links) ? links : [];
    this.fetchedAt = fetchedAt || new Date().toISOString();
    this.format = format;
    this.blocked = Boolean(blocked);
    this.error = error || "";
    this.markdownChars = markdownChars || (markdown ? markdown.length : 0);
    this.success = Boolean(success) && !this.blocked && !this.error;
  }

  toDict() {
    return {
      success: this.success,
      url: this.url,
      title: this.title,
      markdown: this.markdown,
      excerpt: this.excerpt,
      text: this.text,
      links: this.links.map((l) =>
        typeof l === "object"
          ? { href: l.href ?? "", absolute: l.absolute ?? "", text: l.text ?? "" }
          : { href: String(l), absolute: String(l), text: "" },
      ),
      fetched_at: this.fetchedAt,
      format: this.format,
      blocked: this.blocked,
      error: this.error,
      markdown_chars: this.markdownChars,
      text_chars: this.text?.length ?? 0,
    };
  }

  static fromDict(data = {}) {
    return new PageMarkdown({
      url: data.url,
      title: data.title,
      markdown: data.markdown,
      excerpt: data.excerpt,
      text: data.text,
      links: data.links,
      fetchedAt: data.fetched_at ?? data.fetchedAt,
      format: data.format,
      blocked: data.blocked,
      error: data.error,
      markdownChars: data.markdown_chars ?? data.markdownChars,
      success: data.success,
    });
  }

  static fromExploreResult(result, { maxChars = 60000, format = "markdown" } = {}) {
    let markdown = result?.markdown ?? "";
    if (format === "readme") {
      markdown = toReadmeMarkdown({
        title: result?.title,
        url: result?.finalUrl || result?.startUrl,
        markdown,
        links: result?.links,
      });
    }
    markdown = smartTruncate(markdown, maxChars);
    return new PageMarkdown({
      url: result?.finalUrl || result?.startUrl || "",
      title: result?.title || "",
      markdown,
      text: result?.text || "",
      links: result?.links || [],
      format,
      error: result?.success ? "" : result?.error || "fetch failed",
      success: Boolean(result?.success),
      blocked: Boolean(result?.metadata?.blocked),
    });
  }
}

export function makeExcerpt(text, max = 320) {
  const clean = String(text ?? "")
    .replace(/^#+\s.*$/gm, "")
    .replace(/Source:\s+\S+/g, "")
    .replace(/\s+/g, " ")
    .trim();
  if (clean.length <= max) return clean;
  return `${clean.slice(0, max - 3).trimEnd()}...`;
}

export function toReadmeMarkdown({ title = "", url = "", markdown = "", links = [] } = {}) {
  const body = String(markdown ?? "");
  const headings = [...body.matchAll(/^#{1,3}\s+(.+)$/gm)].map((m) => m[1].trim()).slice(0, 12);
  const lines = [];
  lines.push(`# ${title || "Untitled page"}`);
  lines.push("");
  if (url) {
    lines.push(`Source: ${url}`);
    lines.push("");
  }
  if (headings.length) {
    lines.push("## Contents");
    lines.push("");
    for (const h of headings) lines.push(`- ${h}`);
    lines.push("");
  }
  // strip duplicate leading title/source if already present
  let rest = body;
  rest = rest.replace(/^#\s+.*\n+/, "");
  rest = rest.replace(/^Source:\s+\S+\n+/, "");
  lines.push(rest.trim());
  if (links?.length && !rest.includes("## Links")) {
    lines.push("");
    lines.push("## Links");
    lines.push("");
    for (const l of links.slice(0, 40)) {
      const u = l.absolute || l.href || "";
      if (!u) continue;
      lines.push(`- [${l.text || u}](${u})`);
    }
  }
  return `${lines.join("\n").trim()}\n`;
}

File: cpp/packages/handoffkit/src/browser/page.cpp
#include <handoffkit/browser/page.hpp>
#include <handoffkit/browser/util.hpp>

#include <chrono>
#include <regex>
#include <sstream>

namespace handoffkit {
namespace browser {

std::string make_excerpt(std::string_view text, int max_chars) {
    std::string clean(text);
    clean = std::regex_replace(clean, std::regex(R"(^#+\s.*$)", std::regex::multiline), "");
    clean = std::regex_replace(clean, std::regex(R"(Source:\s+\S+)"), "");
    clean = std::regex_replace(clean, std::regex(R"(\s+)"), " ");
    while (!clean.empty() && clean.front() == ' ') clean.erase(clean.begin());
    while (!clean.empty() && clean.back() == ' ') clean.pop_back();
    if (static_cast<int>(clean.size()) <= max_chars) return clean;
    return clean.substr(0, static_cast<std::size_t>(max_chars - 3)) + "...";
}

std::string to_readme_markdown(std::string_view title, std::string_view url,
                               std::string_view markdown, const std::vector<ExtractedLink>& links) {
    std::string body(markdown);
    std::vector<std::string> headings;
    {
        std::regex re(R"(^#{1,3}\s+(.+)$)");
        auto begin = std::sregex_iterator(body.begin(), body.end(), re);
        auto end = std::sregex_iterator();
        for (auto it = begin; it != end && headings.size() < 12; ++it) {
            headings.push_back((*it)[1].str());
        }
    }
    std::ostringstream ss;
    ss << "# " << (title.empty() ? "Untitled page" : title) << "\n\n";
    if (!url.empty()) ss << "Source: " << url << "\n\n";
    if (!headings.empty()) {
        ss << "## Contents\n\n";
        for (const auto& h : headings) ss << "- " << h << "\n";
        ss << "\n";
    }
    body = std::regex_replace(body, std::regex(R"(^#\s+.*\n+)"), "");
    body = std::regex_replace(body, std::regex(R"(^Source:\s+\S+\n+)"), "");
    ss << body;
    if (!links.empty() && body.find("## Links") == std::string::npos) {
        ss << "\n\n## Links\n\n";
        int n = 0;
        for (const auto& l : links) {
            if (n++ >= 40) break;
            const std::string u = l.absolute.empty() ? l.href : l.absolute;
            if (u.empty()) continue;
            ss << "- [" << (l.text.empty() ? u : l.text) << "](" << u << ")\n";
        }
    }
    return ss.str();
}

nlohmann::json PageMarkdown::to_json() const {
    nlohmann::json links_j = nlohmann::json::array();
    for (const auto& l : links) links_j.push_back(l.to_json());
    return {
        {"success", success},
        {"url", url},
        {"title", title},
        {"markdown", markdown},
        {"excerpt", excerpt},
        {"text", text},
        {"links", links_j},
        {"fetched_at", fetched_at},
        {"format", format},
        {"blocked", blocked},
        {"error", error},
        {"markdown_chars", markdown_chars},
        {"text_chars", static_cast<int>(text.size())},
    };
}

PageMarkdown PageMarkdown::from_explore_result(const ExploreResult& result, int max_chars,
                                               std::string_view format) {
    PageMarkdown p;
    p.url = result.final_url.empty() ? result.start_url : result.final_url;
    p.title = result.title;
    p.text = result.text;
    p.links = result.links;
    p.markdown = result.markdown;
    if (format == "readme") {
        p.markdown = to_readme_markdown(p.title, p.url, p.markdown, p.links);
    }
    p.markdown = smart_truncate(p.markdown, max_chars);
    p.excerpt = make_excerpt(p.markdown);
    p.format = std::string(format);
    p.error = result.success ? "" : result.error;
    p.success = result.success;
    p.markdown_chars = static_cast<int>(p.markdown.size());
    const auto now = std::chrono::system_clock::now();
    p.fetched_at = std::to_string(
        std::chrono::duration_cast<std::chrono::seconds>(now.time_since_epoch()).count());
    return p;
}

}  // namespace browser
}  // namespace handoffkit
